### ISSUE TITLE
Arrays are not contextually typed by Iterables even if Iterable is globally defined

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#10237]
From @zakjan on https://github.com/Microsoft/TypeScript/issues/10195#issuecomment-238644633.

A user was using a declaration file for core-js.

```ts
interface IteratorResult<T> {
    done: boolean;
    value?: T;
}

interface Iterator<T> {
    next(value?: any): IteratorResult<T>;
    return?(value?: any): IteratorResult<T>;
    throw?(e?: any): IteratorResult<T>;
}

interface Symbol { }

interface Iterable<T> {
    [Symbol.iterator](): Iterator<T>;
}
interface SymbolConstructor {
    iterator: symbol;
}
declare var Symbol: SymbolConstructor;
interface Array<T> extends Iterable<T> {}

declare function foo(elements: Iterable<[number, string]>): boolean;

foo([[1,"hello"]]);
//  ~~~~~~~~~~~~~ <- Error
```

Expected: No error
Actual: Argument is not compatible (`IteratorResult<(number | string)[]>` not assignable to `IteratorResult<[number, string]>`).

**Comment by mhegazy** [comment]
We only check iterators if target is ES6 or higher. not sure it makes sense if target is ES5 even if you include core-js. for instance, does `for (x of iterable)` be allowed now?

**Comment by kitsonk** [comment]
We (@dojo) had a similar issue and simply added [ArrayLike](https://github.com/dojo/shim/blob/4e73e3eb9334e0381ef1618a27d7779245afed07/src/Map.ts#L39) to our polyfill interfaces that take iterators.  This prevents the syntactical use of `for...of` and we then provide a functional [fill](https://github.com/dojo/shim/blob/4e73e3eb9334e0381ef1618a27d7779245afed07/src/iterator.ts#L118) for that as well.

**Comment by mhegazy** [comment]
closing for now. @DanielRosenwasser please reopen if you have more details.
